### ISSUE TITLE
Mini-batch version

### ISSUE THREAD

**Issue by JerryLead** [rfeinman/pytorch-minimize#15]
Thanks for your great code. I'm now using the L-BFGS optimizer of PyTorch, which can be used for mini-batch training. I'm also trying to use your code in the mini-batch mode. However, after removing the 'loss.backward()' as recommended in the `/train_mnist_Minimizer.py`, I encountered the following warning/error. Is it possible to use the algorithms here in mini-batch mode? Thanks!

```
/datadisk/anaconda3/lib/python3.9/site-packages/torch/autograd/__init__.py:173: UserWarning: Using backward() with 
create_graph=True will create a reference cycle between the parameter and its gradient which can cause a memory leak. We 
recommend using autograd.grad when creating the graph to avoid this. If you have to use this function, make sure to reset the 
.grad fields of your parameters to None after use to break the cycle and avoid the leak. (Triggered internally at  
/opt/conda/conda-bld/pytorch_1646756402876/work/torch/csrc/autograd/engine.cpp:985.)

  Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
```

**Comment by rfeinman** [comment]
Hi @JerryLead

Yes, I get this warning message as well and I don't think it's a concern. Although calling `Tensor.backward(create_graph=True)` is generally bug-prone and not recommended, I have taken all of the necessary precautions here. If you'd like to suggest an alternative implementation that uses `autograd.grad` instead of `Tensor.backward` I'd be open to a PR.

P.S. - deterministic optimizers like L-BFGS are not designed for minibatch training. Although it is conceivable to use them this way, there is very little theory to support these applications. As far as I can tell, most of the tools and tricks that PyTorch uses to make its L-BFGS optimizer support minibatch are pretty ad-hoc. I do not find any literature on them.

**Comment by JerryLead** [comment]
@rfeinman Thanks for your reply. I've also noticed that some researchers/developers have implemented L-BFGS with multiple training methods such as `full-batch`, `full-overlap`, and `multi-batch` at https://github.com/hjmshi/PyTorch-LBFGS. They also provide the referenced papers of these methods. Right now, I'm not sure about how these methods work and their differences. It would be great if you were familiar with them and could tell me. Thanks!

**Comment by rfeinman** [comment]
@JerryLead I had a look at the referenced papers. Indeed it seems there were some studies from the 2010s showing that, with certain modifications, you can use second order methods in the stochastic setting without a blowup. However to my knowledge these papers don’t address the question of whether you _should_ use second order methods with a stochastic objective. The appeal of second order methods is that they provide guarantees about the optimality of the solution given certain assumptions (e.g. convexity, saddle points, etc). For a stochastic objective there are no such guarantees; and furthermore there is no evidence to suggest that second order methods should outperform standard stochastic optimizers like gradient descent w/ momentum or ADAM.

**Comment by JerryLead** [comment]
@rfeinman Thanks for your comments. I've sent my further comments and questions to your NYU email. Looking forward to your further comments.
